### PR TITLE
Add two value logical op

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: ref #9146

Problem Summary:

### What is changed and how it works?
This is a follow up pr for #9523. #9523 optimize logical op in TiFlash, but if the logical op returns a nullable(UInt8) column for  filter executor, in `FilterDescription`, it still need to merge null_map and the data into a new filter vector:
https://github.com/pingcap/tiflash/blob/b36199ea6f2a8ec754ee17db86dd3727e0524b14/dbms/src/Columns/FilterDescription.cpp#L91-L93

This pr add a special logic function called `two_value_and`, and inside `two_value_and`, it will treat null as false, so it always return UInt8 column. This function should only be used for filter executor. In this pr, it only uses `two_value_and` for `PhysicalFilter::build` and `executePushedDownFilter`.

For the same query in #9146 
Orignal time ~0.42s
After #9523 ~0.26s
After this pr ~0.23s 
```commit-message

```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Functions/FunctionsLogical.h [hunk 1]
```diff
@@ -587,13 +654,23 @@ class FunctionAnyArityLogical : public IFunction
...
-                vec_res[i] = static_cast<bool>(col_data[i]);
+                for (size_t i = 0; i < rows; ++i)
+                {
+                    vec_res[i] = !(*null_maps[0])[i] && static_cast<bool>(col_data[i]);
+                }
+            }
+            else
+            {
+                // according to https://github.com/pingcap/tiflash/issues/5849
+                // need to cast the UInt8 column to bool explicitly
```
Comment (gengliqi) [inline dbms/src/Functions/FunctionsLogical.h:667]: These comments can be moved before `if constexpr (null_as_false)`

File: dbms/src/Functions/FunctionsLogical.h [hunk 2]
```diff
@@ -772,24 +874,48 @@ class FunctionAnyArityLogical : public IFunction
...
+                    {
+                        vec_res[i] = static_cast<bool>(col_data[i]);
+                    }
+                    vec_res_is_null.assign(*null_maps[0]);
                 }
-                vec_res_is_null.assign(*null_maps[0]);
             }
             else
             {
+                if (!result_is_nullable)
```
Comment (gengliqi) [inline dbms/src/Functions/FunctionsLogical.h:895]: How about `if constexpr (null_as_false)`?

Comment (windtalker) [reply dbms/src/Functions/FunctionsLogical.h]: Done

File: dbms/src/Functions/FunctionsLogical.h (hunk 1 above)
Comment (windtalker) [reply dbms/src/Functions/FunctionsLogical.h]: Done

**Review by gengliqi — APPROVED** [review]
LGTM

**Review by yibin87 — APPROVED** [review]
LGTM

**Pull request merged**